### DISCUSSION TITLE
Support consolidation of outbox items

### DISCUSSION THREAD

**Discussion by panos-tr** [bluegroundltd/transactional-outbox#14] [Ideas]
## Context / Motivation
The transactional outbox library retrieves all eligible items and calls `OutboxHandler.handle` for each one of them while performing all house keeping (marking them as running/completed/failed, rescheduling them, etc.) internally. This allows for a simple and streamlined implementation on the consumer side since the logic is effectively contained in the `handle` method.

Still, there are cases where it would make sense to provide the ability for consumers to perform complex/batch operations based on multiple items. This includes:
- Using an outbox to asynchronously publish entities. If multiple outbox items exist about the same entity, it makes sense to only take the most recent into account and discard the rest.
- Using an outbox to notify about state changes of an entity. If multiple outbox items exist about the same entity, it might make sense to consolidate all updates into a single one and only process that.

## Suggestion(s)
We should provide a method for consumers to consolidate/group/filter the outbox items being processed.

#### List of items in `handle` method
Update the `handle` method to accept a list of items. While this is straight-forward, it comes with some drawbacks:
1. In order for outbox house keeping to remain in the library we will need to implement complex schemes.
1. It is a breaking change since it changes the exposed `OutboxHandler` interface.

#### Consolidate when adding
Introduce a separate component that could be registered (per outbox type) with the transactional outbox and would run when adding an item. It will have access to all pending outbox items and would be allowed to update them as well as add new ones.

#### Separate consolidation step
Add a pre-process step (probably in the form of a method on `OutboxHandler`) where the required consolidation will be implemented. The method should group the items and provide a representative for each group. The latter will be then passed to the `handle` method.
1. If handling of the representative item succeeds, all `OutboxItem`s in the group should be marked as `COMPLETED`.
a. Alternatively we could use a new status value (e.g. `DEFERRED` or `CONSOLIDATED`).
b. In case of (non-terminal) failure does it make sense to update the status of all items in the group? Or would it be simpler if we only update the status of the representative?
c. We could consider updating the status of non-representative items before `handle` is executed. Then the representative item, effectively becomes the only eligible for execution/retry.
3. Would it be possible to allow the consolidation method to 'build' a new outbox item that would provide a merged view over all of the outbox items in the group?
4. Generally, this could be backwards compatible since it will be easy to provide a default implementation for the consolidation method that does nothing, i.e. builds a separate group for each item with itself as a representative.

**Comment by gerasimosfarantatos** [comment]
A related enhancement would be to add the ability to configure different processing periods per outbox type. That is, instead of all outboxes being processed every 30s (or whatever period the scheduler is configured), some specific items could be processed on a longer period. 

This is related to consolidation as in some use cases there is a bunch of updates on a new entity in the first, say, 5 minutes, and then the entity is updated less frequently. Such updates are usually driven by users and, while the more extended period will never guarantee that we won't end up with multiple invocations of the handlers, it will mitigate the issue.

As far as the implementation is concerned, I suppose we can find a way of keeping the execution interval of the scheduler unchanged and play around with the execution timestamps.

**Comment by gerasimosfarantatos** [comment]
> Separate consolidation step

I think I like this one among the alternatives

**Comment by kgiannakelos** [comment]
Possible case here as pipointed by @panos-tr https://github.com/bluegroundltd/atlas/pull/8436#discussion_r1170219868
